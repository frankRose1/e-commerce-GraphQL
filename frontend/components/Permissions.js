import React from 'react';
import {Query} from 'react-apollo';
import gql from 'graphql-tag';
import Error from './ErrorMessage';
import Table from './styles/Table';
import UserPermissions from './UserPermissions';

const possiblePermissions = [
  'ADMIN',
  'USER',
  'ITEMCREATE',
  'ITEMUPDATE',
  'ITEMDELETE',
  'PERMISSIONUPDATE'
];

const ALL_USERS_QUERY = gql`
  query ALL_USERS_QUERY {
    usersPermissions {
      id
      name
      email
      permissions
    }
  }
`;

const Permissions = () => (
  <Query query={ALL_USERS_QUERY}>
    {({data, loading, error}) => (
      <div>
        <Error error={error}/>
        <div>
          <h2>Manage Permissions</h2>
          <Table>
            <thead>
              <tr>
                <th>Name</th>
                <th>Email</th>
                {possiblePermissions.map((perm, i) => (
                  <th key={i}>{perm}</th>
                ))}
                <th>Update</th>
              </tr>
            </thead>
            <tbody>
              {data.usersPermissions.map(user => (
                <UserPermissions
                  key={user.id}
                  possiblePermissions={possiblePermissions}
                  user={user}/>
              ))}
            </tbody>
          </Table>
        </div>
      </div>
    )}
  </Query>
);

export default Permissions;
export {ALL_USERS_QUERY};