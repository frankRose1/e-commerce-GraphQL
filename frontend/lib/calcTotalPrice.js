/**
 * Take in a users cart and return the total cost of all items in cart
 * @param {array} cart - array of cartItems which contain data about a specific item in the DB
 * @return {integer} totalPrice - total of cart items
 */
function calcTotalPrice(cart){
  return cart.reduce((tally, cartItem) => {
    //in the case an item is deleted from the DB the cartItem will have an ID that points to a deleted item
    if (!cartItem.item) return tally;
    return tally + (cartItem.quantity * cartItem.item.price);
  }, 0);
}

export default calcTotalPrice;