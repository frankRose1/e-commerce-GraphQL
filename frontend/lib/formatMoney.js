/**
 * Format dollar amount of an item
 * @param {integer} amount - dollar amount of an item in cents
 * @return newly formatted price as a string, US currency 
 */
export default function(amount){
  const options = {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2
  }
  //for whole dollar amounts, leave off the .00
  if (amount % 100 === 0) options.minimumFractionDigits = 0;
  const formatter = new Intl.NumberFormat('en-US', options);
  return formatter.format(amount / 100);
}