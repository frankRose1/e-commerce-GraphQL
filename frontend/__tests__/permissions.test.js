import { mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';
import toJSON from 'enzyme-to-json';
import Permissions, { ALL_USERS_QUERY } from '../components/Permissions';
import wait from 'waait';
import { fakeUser } from '../lib/testUtils';

const mocks = [
  {
    request: { query: ALL_USERS_QUERY },
    result: {
      data: {
        usersPermissions: [
          fakeUser(),
          {
            ...fakeUser(),
            id: '5467',
            permissions: ['USER']
          }
        ]
      }
    }
  }
];

describe('<Permissions />', () => {
  it('renders and matches the snapshot', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Permissions />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const table = wrapper.find('table[data-test="permissions-table"]');
    expect(toJSON(table)).toMatchSnapshot();
  });

  it('renders a table row for each user', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Permissions />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const table = wrapper.find('table[data-test="permissions-table"]');
    expect(table.find('tbody').children()).toHaveLength(2);
  });
});
