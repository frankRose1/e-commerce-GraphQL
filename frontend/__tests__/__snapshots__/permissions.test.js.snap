// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Permissions /> renders and matches the snapshot 1`] = `
<table
  className="Table-ajjtbc-0 fctRqQ"
  data-test="permissions-table"
>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Email
      </th>
      <th
        key="0"
      >
        ADMIN
      </th>
      <th
        key="1"
      >
        USER
      </th>
      <th
        key="2"
      >
        ITEMCREATE
      </th>
      <th
        key="3"
      >
        ITEMUPDATE
      </th>
      <th
        key="4"
      >
        ITEMDELETE
      </th>
      <th
        key="5"
      >
        PERMISSIONUPDATE
      </th>
      <th>
        Update
      </th>
    </tr>
  </thead>
  <tbody>
    <UserPermissions
      key="4234"
      possiblePermissions={
        Array [
          "ADMIN",
          "USER",
          "ITEMCREATE",
          "ITEMUPDATE",
          "ITEMDELETE",
          "PERMISSIONUPDATE",
        ]
      }
      user={
        Object {
          "__typename": "User",
          "email": "Delmer.Smith@yahoo.com",
          "id": "4234",
          "name": "Miss Coleman Berge",
          "permissions": Array [
            "ADMIN",
          ],
          Symbol(id): "User:4234",
        }
      }
    >
      <Mutation
        mutation={
          Object {
            "definitions": Array [
              Object {
                "directives": Array [],
                "kind": "OperationDefinition",
                "name": Object {
                  "kind": "Name",
                  "value": "UPDATE_PERMISSIONS_MUTATION",
                },
                "operation": "mutation",
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "userId",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "userId",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "permissions",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "permissions",
                            },
                          },
                        },
                      ],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "updateUserPermissions",
                      },
                      "selectionSet": Object {
                        "kind": "SelectionSet",
                        "selections": Array [
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "id",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "name",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "email",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "permissions",
                            },
                            "selectionSet": undefined,
                          },
                        ],
                      },
                    },
                  ],
                },
                "variableDefinitions": Array [
                  Object {
                    "defaultValue": undefined,
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "ID",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "userId",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "ListType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "Permission",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "permissions",
                      },
                    },
                  },
                ],
              },
            ],
            "kind": "Document",
            "loc": Object {
              "end": 215,
              "start": 0,
            },
          }
        }
        variables={
          Object {
            "permissions": Array [
              "ADMIN",
            ],
            "userId": "4234",
          }
        }
      >
        <tr>
          <td>
            Miss Coleman Berge
          </td>
          <td>
            Delmer.Smith@yahoo.com
          </td>
          <td
            key="0"
          >
            <label
              htmlFor="4234-permission-ADMIN"
            >
              <input
                checked={true}
                disabled={false}
                id="4234-permission-ADMIN"
                onChange={[Function]}
                type="checkbox"
                value="ADMIN"
              />
            </label>
          </td>
          <td
            key="1"
          >
            <label
              htmlFor="4234-permission-USER"
            >
              <input
                checked={false}
                disabled={false}
                id="4234-permission-USER"
                onChange={[Function]}
                type="checkbox"
                value="USER"
              />
            </label>
          </td>
          <td
            key="2"
          >
            <label
              htmlFor="4234-permission-ITEMCREATE"
            >
              <input
                checked={false}
                disabled={false}
                id="4234-permission-ITEMCREATE"
                onChange={[Function]}
                type="checkbox"
                value="ITEMCREATE"
              />
            </label>
          </td>
          <td
            key="3"
          >
            <label
              htmlFor="4234-permission-ITEMUPDATE"
            >
              <input
                checked={false}
                disabled={false}
                id="4234-permission-ITEMUPDATE"
                onChange={[Function]}
                type="checkbox"
                value="ITEMUPDATE"
              />
            </label>
          </td>
          <td
            key="4"
          >
            <label
              htmlFor="4234-permission-ITEMDELETE"
            >
              <input
                checked={false}
                disabled={false}
                id="4234-permission-ITEMDELETE"
                onChange={[Function]}
                type="checkbox"
                value="ITEMDELETE"
              />
            </label>
          </td>
          <td
            key="5"
          >
            <label
              htmlFor="4234-permission-PERMISSIONUPDATE"
            >
              <input
                checked={false}
                disabled={false}
                id="4234-permission-PERMISSIONUPDATE"
                onChange={[Function]}
                type="checkbox"
                value="PERMISSIONUPDATE"
              />
            </label>
          </td>
          <td>
            <SickButton
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <button
                className="SickButton-epeo8a-0 hDvtOo"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Updat
                e
              </button>
            </SickButton>
          </td>
        </tr>
      </Mutation>
    </UserPermissions>
    <UserPermissions
      key="5467"
      possiblePermissions={
        Array [
          "ADMIN",
          "USER",
          "ITEMCREATE",
          "ITEMUPDATE",
          "ITEMDELETE",
          "PERMISSIONUPDATE",
        ]
      }
      user={
        Object {
          "__typename": "User",
          "email": "Gilbert.Nicolas@yahoo.com",
          "id": "5467",
          "name": "Ms. Nigel Russel",
          "permissions": Array [
            "USER",
          ],
          Symbol(id): "User:5467",
        }
      }
    >
      <Mutation
        mutation={
          Object {
            "definitions": Array [
              Object {
                "directives": Array [],
                "kind": "OperationDefinition",
                "name": Object {
                  "kind": "Name",
                  "value": "UPDATE_PERMISSIONS_MUTATION",
                },
                "operation": "mutation",
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "userId",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "userId",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "permissions",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "permissions",
                            },
                          },
                        },
                      ],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "updateUserPermissions",
                      },
                      "selectionSet": Object {
                        "kind": "SelectionSet",
                        "selections": Array [
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "id",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "name",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "email",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "permissions",
                            },
                            "selectionSet": undefined,
                          },
                        ],
                      },
                    },
                  ],
                },
                "variableDefinitions": Array [
                  Object {
                    "defaultValue": undefined,
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "NonNullType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "ID",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "userId",
                      },
                    },
                  },
                  Object {
                    "defaultValue": undefined,
                    "kind": "VariableDefinition",
                    "type": Object {
                      "kind": "ListType",
                      "type": Object {
                        "kind": "NamedType",
                        "name": Object {
                          "kind": "Name",
                          "value": "Permission",
                        },
                      },
                    },
                    "variable": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "permissions",
                      },
                    },
                  },
                ],
              },
            ],
            "kind": "Document",
            "loc": Object {
              "end": 215,
              "start": 0,
            },
          }
        }
        variables={
          Object {
            "permissions": Array [
              "USER",
            ],
            "userId": "5467",
          }
        }
      >
        <tr>
          <td>
            Ms. Nigel Russel
          </td>
          <td>
            Gilbert.Nicolas@yahoo.com
          </td>
          <td
            key="0"
          >
            <label
              htmlFor="5467-permission-ADMIN"
            >
              <input
                checked={false}
                disabled={false}
                id="5467-permission-ADMIN"
                onChange={[Function]}
                type="checkbox"
                value="ADMIN"
              />
            </label>
          </td>
          <td
            key="1"
          >
            <label
              htmlFor="5467-permission-USER"
            >
              <input
                checked={true}
                disabled={false}
                id="5467-permission-USER"
                onChange={[Function]}
                type="checkbox"
                value="USER"
              />
            </label>
          </td>
          <td
            key="2"
          >
            <label
              htmlFor="5467-permission-ITEMCREATE"
            >
              <input
                checked={false}
                disabled={false}
                id="5467-permission-ITEMCREATE"
                onChange={[Function]}
                type="checkbox"
                value="ITEMCREATE"
              />
            </label>
          </td>
          <td
            key="3"
          >
            <label
              htmlFor="5467-permission-ITEMUPDATE"
            >
              <input
                checked={false}
                disabled={false}
                id="5467-permission-ITEMUPDATE"
                onChange={[Function]}
                type="checkbox"
                value="ITEMUPDATE"
              />
            </label>
          </td>
          <td
            key="4"
          >
            <label
              htmlFor="5467-permission-ITEMDELETE"
            >
              <input
                checked={false}
                disabled={false}
                id="5467-permission-ITEMDELETE"
                onChange={[Function]}
                type="checkbox"
                value="ITEMDELETE"
              />
            </label>
          </td>
          <td
            key="5"
          >
            <label
              htmlFor="5467-permission-PERMISSIONUPDATE"
            >
              <input
                checked={false}
                disabled={false}
                id="5467-permission-PERMISSIONUPDATE"
                onChange={[Function]}
                type="checkbox"
                value="PERMISSIONUPDATE"
              />
            </label>
          </td>
          <td>
            <SickButton
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              <button
                className="SickButton-epeo8a-0 hDvtOo"
                disabled={false}
                onClick={[Function]}
                type="button"
              >
                Updat
                e
              </button>
            </SickButton>
          </td>
        </tr>
      </Mutation>
    </UserPermissions>
  </tbody>
</table>
`;
